### PR TITLE
Add unit test coverage for CycleCountService

### PR DESCRIPTION
**Pull request by christianallen**
## Description

Per discussion, this adds a unit spec for `CycleCountService` as a first sample of the test coverage work we would like to contribute more broadly. The goal was to show the approach: cover the service's behavior and its edge cases with assertions that fail when the logic changes, using the test infrastructure the project already has.

`CycleCountServiceSpec` - 63 tests, Spock + `DataTest`, same conventions as the existing `CycleCountTransactionServiceSpec`. The service's two collaborators are mocked: `CycleCountProductAvailabilityService` (the "stock records" in the outline below) and `CycleCountTransactionService`. The outline lists what each test proves, grouped by the service method under test:

**Getting the item count for a count request** - `getInventoryItemsCount(CycleCountRequest)`

1. When a count has not started yet, the item count comes from the current stock records.
2. When stock records return nothing, the item count is null instead of crashing.
3. Once a count has started, only the latest round's items are counted, and stock records are not queried again.

**Creating and updating count requests** - `createRequests(CycleCountRequestBatchCommand)` for 4-5, `updateRequests(CycleCountRequestUpdateBulkCommand)` for 6-9

4. Creating requests saves one per row, marked as a manual request, with each row's own blind-count setting.
5. Creation stops at the first invalid request and reports the validation error.
6. An assignment for the first count sets the count assignee and deadline, and nothing else.
7. An assignment for the recount sets the recount assignee and deadline, and nothing else.
8. An empty update leaves existing assignments alone.
9. Updating a request that fails validation raises an error instead of silently skipping the save.

**Starting a count** - `startCycleCount(CycleCountStartCommand, Location)` for 10-11 (11 also exercises `updateCountAssigneeData`), `startCycleCount(CycleCountStartBatchCommand)` for 12

10. A new count gets one ready-to-count line per stock record, with quantity and bin copied over, zero-stock lots pre-filled with a counted quantity of 0, in-stock lots left blank, the assignee applied, and the request linked and marked in progress.
11. Starting an already-started count is safe: no new lines, no stock lookup, and the first-count assignees are refreshed while recount lines keep theirs.
12. Starting a batch of requests creates one count per request.

**Starting a recount** - `startRecount(Location, CycleCountStartRecountCommand)` (14 also exercises `updateRecountAssigneeData`)

13. A fresh recount builds its lines from up-to-date stock records (including lots that appeared after the first count), assigned to the recounter, without touching the first count.
14. Starting an already-started recount is safe: no new lines, and only recount assignees are refreshed.
15. A line the user added by hand during the count is carried into the recount as a normal line, uncounted, and (unlike the other lines) without an assignee.
16. A hand-added line whose lot now shows up in stock records is not duplicated in the recount.

**Submitting a count** - `submitCount(CycleCountSubmitCountCommand)` (20-21 go through `determineCycleCountItemStatusForSubmit`, 22-25 through `closeCycleCount`, with `CycleCountTransactionService.createTransactions` mocked and verified by interaction)

17. Submitting fails with a clear error when quantities changed underneath the count and the caller asked to be strict about it.
18. Submitting proceeds when quantities changed but the caller did not ask to be strict.
19. The stock-record refresh is skipped entirely unless the caller asks for it.
20. When recounts are required, a line whose counted quantity does not match is held back for recount.
21. When recounts are not required, a mismatched line is approved anyway and the count completes.
22. A count held back for recount posts no inventory transactions and the request stays in progress.
23. A fully resolved count posts its transactions exactly once and the request completes.
24. Products whose lines get dropped by the refresh still get their baseline transaction, because the product list is snapshotted before the refresh (guards the OBPIH-7525 fix).
25. A count whose lines were all canceled still completes and posts transactions (discussion item 1 below).

**Updating count lines** - `updateCycleCountItem(CycleCountUpdateItemCommand)` for 26-29, `updateCycleCountItems(List<CycleCountUpdateItemCommand>)` for 30

26. Saving a line during the count puts it in the first count with a "counting" status and updates the overall count status.
27. Saving a line during the recount puts it in the recount with an "investigating" status and updates the overall count status.
28. A count date supplied with the update is kept.
29. An update without a count date stamps the current time.
30. A batch of line updates applies every one and returns them in the same order.

**Adding lines by hand** - `createCycleCountItem(CycleCountItemCommand)` for 31-35, `createCycleCountItems(List<CycleCountItemCommand>)` for 36

31. A hand-added line starts from zero stock on hand, joins the current count, and reopens a completed count.
32. A hand-added line during a recount goes into the recount with an "investigating" status.
33. Entering a product and lot that already exist reuses the existing inventory record instead of creating a duplicate.
34. Entering a genuinely new lot creates and saves its inventory record once.
35. A discrepancy reason that is not on the cycle-count list is rejected.
36. Two rows in one batch entering the same new lot end up sharing a single inventory record.

**Deleting lines** - `deleteCycleCountItem(String)`

37. A hand-added line can be deleted, and its neighbors survive.
38. A system-generated line refuses deletion with a clear error.
39. Deleting an unknown line id does nothing rather than erroring (documents current behavior).

**Deleting requests** - `deleteCycleCountRequests(List<String>)` (private `deleteCycleCountRequest` and `deleteCycleCount` underneath)

40. Deleting a request also deletes its count and all its lines.
41. Deleting a request that never started a count works cleanly.
42. Deleting an unknown request id fails with a clear not-found error.

**Deleting a count and everything attached to it** - `deleteCycleCountWithAssociations(String)` (private `deleteCycleCountTransactions` underneath)

43. Removes the count, its lines, its request, and the inventory transactions the count created, leaving nothing behind.
44. An unknown count id fails with a clear not-found error.

**Refreshing a count against current stock** - `refreshCycleCount(String, boolean, Integer)` (49-52 verify the delegation to the mocked `CycleCountProductAvailabilityService`)

45. Refreshing an unknown count id fails with a clear not-found error.
46. A completed count refuses to refresh, with the exact error message.
47. A canceled count refuses to refresh.
48. A count awaiting review refuses to refresh.
49. A just-requested count refreshes, passing the caller's options through untouched.
50. A count in progress refreshes the same way.
51. A count with recorded quantities refreshes the same way.
52. A count under investigation refreshes the same way.

**Exporting the candidate list as CSV** - `getCycleCountCsv(List<CycleCountCandidate>)`

53. A fully populated candidate renders the exact expected header and data row, so a swapped or dropped column cannot slip through (also documents that comma-containing values get escaped twice).
54. A candidate with almost no data renders empty cells for missing text and 0 for missing quantity.
55. An empty candidate list renders just the header.
56. A null candidate list also renders just the header.

**Exporting the count form** - `getCountFormXls(List<CycleCountDto>)`

57. Each line renders one spreadsheet row with lot, expiration date, bin, assignee, and dates formatted as expected.
58. Uncounted lines and missing optional fields render as blanks, while a counted quantity of zero renders as 0, because blank means "not counted yet".

**Exporting the recount form** - `getRecountFormXls(List<CycleCountDto>)` (private `getRecountAsXlsMap` and `mergeCountAndRecountItemAsXlsMap` underneath)

59. A first count and its recount of the same lot and bin merge into a single row showing both quantities, the difference, and both dates.
60. A difference of exactly zero renders as a blank cell (discussion item 2 below).
61. A first-count line whose lot vanished from stock before the recount currently crashes the export (discussion item 3 below).
62. A hand-added recount line gets its own row with recount data and empty first-count columns.
63. A recounted quantity of zero stays visible as 0 rather than blank.

The criteria-query methods (`getCandidates`, `getPendingCycleCountRequests`, `getCycleCounts`, and the two report queries) are intentionally not unit-tested here - they belong to integration-level coverage.

**Three behaviors for discussion, found while writing these tests**, flagged here for now, bringing back for discussion before proceeding / fixing:

1. A count whose items were all CANCELED resolves to COMPLETED (`recomputeStatus` treats canceled items as completed), so the cancellation branch of the close-out logic is unreachable from `submitCount` and transactions are still committed for a fully canceled count (test 25). Should a fully canceled count end up CANCELED instead?
2. A count difference of exactly 0 renders as blank in the recount export ("Difference" goes through a truthiness check), which is hard to distinguish from "not counted" (test 60). Is the blank intentional?
3. `getRecountFormXls` throws a `NullPointerException` when an initial-count item has no recount counterpart (no pun intended). This happens when a lot disappears from availability between count and recount (test 61). The test documents the current behavior.

## Tests

Test-only change; no production code touched. `./gradlew test --tests "org.pih.warehouse.inventory.CycleCountServiceSpec"` - 63 tests green, and the full cycle-count spec family (104 tests) passes together. Backend suite run against both MySQL 8.0.36 and MariaDB 10.3.39, frontend suite and lint unaffected but run as well.

### CODE REVIEW

**Comment by ewaterman** [comment]
sorry I haven't gotten around to reviewing this yet. We're quite pressed on wrapping up the 0.9.9 release right now. I'll try to find time for this next week

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,1536 @@
...
+                quantityCounted: args.containsKey("quantityCounted") ? args.quantityCounted : null,
+                custom: args.containsKey("custom") ? args.custom : false,
+                assignee: args.assignee,
+                dateCounted: args.dateCounted,
+        )
+    }
+
+    private AvailableItem availableItem(InventoryItem inventoryItem, Location binLocation, BigDecimal quantityOnHand) {
+        return new AvailableItem(inventoryItem: inventoryItem, binLocation: binLocation, quantityOnHand: quantityOnHand)
+    }
```
Comment (ewaterman) [inline src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:161]: All the methods you have here are simple object initializers with minimal logic, so I'm fine with it, but I just want to hilight that in general I don't think we should let our tests get too "smart". The more logic we put in our tests, the more we get stuck in a "testing the tests" scenario. I'm generally in favour of tests being as verbose as they need, and that some amount of code duplication is fine.

but again, what you've done here is fine so no changes required

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,1536 @@
...
+        CycleCountRequest request = createRequest(facility: facility, product: product, cycleCount: cycleCount)
+
+        when:
+        Integer count = cycleCountService.getInventoryItemsCount(request)
+
+        then: 'the recount (count index 1) is the most recent count, and it has a single item'
+        assert count == 1
+
+        and: 'availability is never consulted once a count exists'
+        0 * cycleCountProductAvailabilityServiceMock.getAvailableItems(*_)
```
Comment (ewaterman) [inline src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:215]: do we care if we query availability or not? I don't actually remember the behaviour here. If not calling available items is an important piece of the functionality, then I agree with this assert, but if it is more of a side-effect, then I don't think there's a need to make assertions on it. We should mainly care about the inputs and outputs of a method and treat its internals as a black box (unless we need to stub external calls that it makes).

But again, if the fact that it does not refresh availability is important to the implementation, then this assert is fine

EDIT: I dug into the code a bit. I'm convinced that not refreshing QoH is important enough to the flow to include the assert

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,1536 @@
...
+        when:
+        cycleCountService.createRequests(command)
+
+        then: 'the batch fails before returning anything'
+        thrown(ValidationException)
+        // The service validates and saves one request at a time, so the valid first request has
+        // already been saved when the second one fails. Whether that first save actually reaches
+        // the database is up to the surrounding transaction's rollback, which a unit test
+        // does not cover.
+    }
```
[the remark this answers is not in the document]
Comment (ewaterman) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy]: I think I agree with Copilot here.

Unit tests should not test the db layer, but you're explicitly stating that you're testing that it "stops at the first invalid request", which to do you need to check the db. IMO interacting with the db in this way is fine to do since you're mocking the db interaction and are already persisting data to this mocked db (via save() calls) but if it makes you uncomfortable to do, I just wouldn't bother testing for that scenario and simply change the test name to reflect that all you're doing is expecting a failure if any item is invalid

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,1536 @@
...
+        assert dto.id == cycleCount.id
+
+        where:
+        status << [CycleCountStatus.REQUESTED, CycleCountStatus.COUNTING,
+                   CycleCountStatus.COUNTED, CycleCountStatus.INVESTIGATING]
+    }
+
+    // ---------------------------------------------------------------------------------------------
+    // Export: CSV of cycle count candidates
+    // ---------------------------------------------------------------------------------------------
```
Comment (ewaterman) [inline src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:1347]: I haven't looked too closely at the exporter tests. As an fyi, I'm in the process of finishing a refactor of how our exporters work in general. Eventually all of the export logic will be pulled out from cycle count and instead exports will be performed by specifying the desired content type on a list endpoint, which then calls into general purpose exporter components for serialization (which have their own tests).

If you're curious about how that is implemented, you can see this PR stack: https://github.com/openboxes/openboxes/pull/6103

Anyways, I don't expect to have this fully integrated with cycle count any time soon so having these tests for now is nice

**Review by ewaterman — COMMENTED** [review]
I haven't had a chance to look over everything closely, but I wanted to at least get some comments in for you since you've been waiting a long time. In general it all looks good to me 👍

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,1536 @@
...
+        User currentUser = new User(username: "tester")
+        currentUser.save(validate: false, flush: true)
+        new AuthService().setCurrentUser(currentUser)
+
+        cycleCountProductAvailabilityServiceMock = Mock(CycleCountProductAvailabilityService)
+        cycleCountService.cycleCountProductAvailabilityService = cycleCountProductAvailabilityServiceMock
+
+        cycleCountTransactionServiceMock = Mock(CycleCountTransactionService)
+        cycleCountService.cycleCountTransactionService = cycleCountTransactionServiceMock
+    }
```
[the remark this answers is not in the document]
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:66]: Added a cleanup() that clears it. d6f808f3f

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy]: Went with the rename. The test now claims only that an invalid request fails the batch. I tried the persisted-state check first, but after the failed validation the pending first save never flushes in the unit environment, so the count came back 0 - which says nothing about what the transaction does in production. d6f808f3f

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,1536 @@
...
+        assert carried.quantityCounted == null
+        assert carried.quantityOnHand == 0
+        assert carried.status == CycleCountItemStatus.INVESTIGATING
+
+        and: 'the carried item is intentionally no longer custom, so it cannot be deleted from the recount'
+        assert !carried.custom
+
+        and: 'the recount assignee is only applied to availability-backed items, not carried custom ones'
+        assert recountItems.find { it.inventoryItem == trackedLot }.assignee == recounter
+        assert carried.assignee == null
```
[the remark this answers is not in the document]
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:589]: Pinned it. The carry-forward test now restarts the recount and asserts the carried item picks up the assignee. Whether carried rows should be assigned from the start is a fair question, listed in the description. d6f808f3f

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,1536 @@
...
+                { List<Product> products ->
+                    products.size() == 2 && products.toSet() == [productKept, productRemoved].toSet()
+                }, true)
+    }
+
+    void 'submitCount should treat a count whose items were all canceled as completed'() {
+        // Documents today's behavior: recomputeStatus() treats canceled items as resolved, so a
+        // count whose items were ALL canceled comes out COMPLETED, never CANCELED. As a result the
+        // cancellation branch of the close-out logic can never run from submitCount, and a fully
+        // canceled count still creates inventory transactions for its products.
```
[the remark this answers is not in the document]
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy]: Correct - the item is re-derived to APPROVED before the recompute. Comment rewritten. The overwrite is now asserted, and the description updated to match. d6f808f3f

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,1536 @@
...
+                binLocation: createBinLocation("A1"), countIndex: 0)
+        createCycleCountWithItems(facility, [item])
+        item.save(validate: false, flush: true)
+
+        when:
+        cycleCountService.updateCycleCountItem(new CycleCountUpdateItemCommand(cycleCountItem: item, recount: false))
+
+        then:
+        assert item.dateCounted != null
+        assert Math.abs(item.dateCounted.time - new Date().time) < 5000
```
[the remark this answers is not in the document]
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy]: Replaced with bounds captured around the call. d6f808f3f

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,1536 @@
...
+        given:
+        Category category = new Category(name: "Medicines")
+        category.save(validate: false)
+        Product product = createProduct("AB12", "Ibuprofen 200mg")
+        product.category = category
+        CycleCountCandidate candidate = new CycleCountCandidate(
+                product: product,
+                abcClass: "A",
+                internalLocations: "A1,B2",
+                dateLastCount: new Date(1200000000000L),
```
[the remark this answers is not in the document]
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy]: Fixture is now a parsed calendar date. d6f808f3f

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,1536 @@
...
+        List<Map> rows = cycleCountService.getCountFormXls([CycleCountDto.toDto(cycleCount)])
+
+        then:
+        assert rows.size() == 1
+        assert rows[0]["Product Code"] == "AB12"
+        assert rows[0]["Product Name"] == "Ibuprofen 200mg"
+        assert rows[0]["Lot Number"] == "LOT-1"
+        assert rows[0]["Expiration Date"] == "01/31/2027"
+        assert rows[0]["Bin Location"] == "A1"
+        assert rows[0]["User Counted"] == assignee.name
```
[the remark this answers is not in the document]
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:1425]: Added the assertion. d6f808f3f

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy (hunk 1 above)
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:161]: Agreed, and noted for the specs that follow - the helpers stay plain constructors.

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy (hunk 2 above)
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:215]: Same conclusion: a started count re-querying stock would silently change the item count mid-count, so the assert is part of the contract.

File: src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy (hunk 4 above)
Comment (christianallen) [reply src/test/groovy/org/pih/warehouse/inventory/CycleCountServiceSpec.groovy:1347]: Noted - happy for the export tests to be dropped or moved when the #6103 stack reaches cycle count.

**Comment by christianallen** [comment]
Addressed every thread in d6f808f3f (63 tests, still green on both databases). Description updated: discussion item 1 corrected, item 4 added.

**Comment by christianallen** [comment]
The failures on the last push were upstream drift: develop merged the DTO mapper components (#6088) after this branch was cut, so the count DTO could no longer be built in the unit-test context. Merged develop in and adapted the spec in 845fdcfb0 - the smartMapper bean is now defined in the spec with the real item and product mappers behind it - and added a test for the new no-stock guard on startCycleCount. 64 tests green.
